src/main/java/org/assertj/core/internal/ComparatorBasedComparisonStrategy.java

```diff
@@ -94,6 +95,10 @@ public boolean iterableContains(Iterable<?> iterable, Object value) {
   @SuppressWarnings("unchecked")
   public void iterableRemoves(Iterable<?> iterable, Object value) {
     if (iterable == null) return;
+    // Avoid O(N^2) complexity of serial removal from an iterator of collections like ArrayList
+    if (iterable instanceof Collection) {
+      ((Collection<?>) iterable).removeIf(o -> comparator.compare(o, value) == 0);
```

Comment:
the javadoc of `removeIf` states:
> The default implementation traverses all elements of the collection using its iterator. Each matching element is removed using Iterator.remove(). If the collection's iterator does not support removal then an UnsupportedOperationException will be thrown on the first matching element.

this is what we do here below so before accepting these changes I would like to see before and after performance measurements that would prove that performance was indeed improved (not as an automated test, just as a few perf run).

Reply:
Please check out `ArrayList.removeIf()` implementation, it overrides the default impl.

You can try to comment out the changes in the Strategy and run the `ContainsOnlyPerfTest` without a time limit. I don't how much it takes to complete. I don't have enough patience, after waiting for more than a few dozen seconds. Its complexity is O(1000 billion).

Reply:
I have run `ContainsOnlyPerfTest` with and without the changes, it is indeed much faster, I still fail to see where to is the O(n2), enlighten me please ;-)

Reply:
`ArrayList$Iterator.remove()`, while at the element at index 0, should shift all elements until the end (say there is 1 million of them) by one array slot:

https://github.com/openjdk/jdk/blob/36bd09dead31636e71ab9d74edff130cbd5124a4/src/java.base/share/classes/java/util/ArrayList.java#L636-L642

`System.arrayCopy()` is an O(N) operation here

Reply:
ok I got it I think in `ArrayList.removeIf`:
```
        // Tolerate predicates that reentrantly access the collection for
        // read (but writers still get CME), so traverse once to find
        // elements to delete, a second pass to physically expunge.
```